### PR TITLE
Add option to sync metadata in k8sclusterreceiver

### PR DESCRIPTION
**Pull request by asuresh4**
- Improve metadata sync. Added `KubernetesMetadataUpdate` which provides a delta view of properties on a resource between k8s update events. This will struct explicitly provides properties to be added, removed and updated. This struct will be exposed to exporters.
- Add `KubernetesMetadataExporter` interface
- Add `metadata_exporters` option to sync metadata. Any exporter that implements `KubernetesMetadataExporter` can be specified as a metadata exporter.

Unblocks https://github.com/open-telemetry/opentelemetry-collector-contrib/issues/180

### CODE REVIEW

File: receiver/k8sclusterreceiver/watcher.go
Comment (jrcamp) [inline receiver/k8sclusterreceiver/watcher.go:134]:
```suggestion
func (rw *resourceWatcher) setupMetadataExporters(
	exporters map[configmodels.DataType]map[configmodels.Exporter]component.Exporter,
	metadataExportersFromConfig []string
) error {
```

File: receiver/k8sclusterreceiver/watcher.go
Comment (jrcamp) [inline receiver/k8sclusterreceiver/watcher.go:133]: Do you need the double exporters map here or can you just pass around `exporters[configmodels.MetricsDataType]`?

File: receiver/k8sclusterreceiver/watcher.go
Comment (jrcamp) [inline receiver/k8sclusterreceiver/watcher.go:166]: Between multiple consumers or async to consumers themselves?

File: receiver/k8sclusterreceiver/mock_exporter.go
Comment (jrcamp) [inline receiver/k8sclusterreceiver/mock_exporter.go:1]: If this is only need for tests might be better to end it in _test so it only gets built then.

File: receiver/k8sclusterreceiver/watcher.go
Comment (jrcamp) [inline receiver/k8sclusterreceiver/watcher.go:140]: I think metadata_exporter config values should reference exporters by name instead of type, in case you have multiple and want to only send to one or the other.

File: receiver/k8sclusterreceiver/watcher.go
Comment (asuresh4) [reply receiver/k8sclusterreceiver/watcher.go:166]: Between multiple consumers.

File: receiver/k8sclusterreceiver/watcher.go
Comment (asuresh4) [reply receiver/k8sclusterreceiver/watcher.go:133]: Yes, that's a good point. We can extend later on to support metadata sync in other pipeline types when/if this receiver supports other types.

File: receiver/k8sclusterreceiver/watcher.go
Comment (tigrannajaryan) [inline receiver/k8sclusterreceiver/watcher.go:165]: Is this TODO needed? Is there any indication that current approach is not good enough?

File: receiver/k8sclusterreceiver/README.md
Comment (tigrannajaryan) [inline receiver/k8sclusterreceiver/README.md:66]: It would be nice to also add link to the source code that has more details about the interface.

File: receiver/k8sclusterreceiver/collection/metadata.go
Comment (asuresh4) [reply receiver/k8sclusterreceiver/collection/metadata.go:34]: Added comments

File: receiver/k8sclusterreceiver/collection/metadata.go
Comment (asuresh4) [reply receiver/k8sclusterreceiver/collection/metadata.go]: Added comments on the method. Also, updated this method to take in a slice instead of map since we only really care about the updates themselves and don't need a mapping.

File: receiver/k8sclusterreceiver/watcher.go
Comment (asuresh4) [reply receiver/k8sclusterreceiver/watcher.go:156]: Added `validateMetadataExporters` method to perform this validation.

File: receiver/k8sclusterreceiver/watcher.go
Comment (asuresh4) [reply receiver/k8sclusterreceiver/watcher.go:186]: I was thinking of cases where there's more than one consumers of the metadata. In that case if the method being invoked blocks, performing these calls synchronously would slow down the metadata sync for other exporters. What do you think?

File: receiver/k8sclusterreceiver/README.md
Comment (asuresh4) [reply receiver/k8sclusterreceiver/README.md:66]: Done.

File: receiver/k8sclusterreceiver/collection/metadata.go
Comment (tigrannajaryan) [inline receiver/k8sclusterreceiver/collection/metadata.go:153]: 3 maps of the same type in the return are very easy to confuse the order. It will be nicer to declare a struct with named fields and return the struct by value instead.

File: receiver/k8sclusterreceiver/collection/metadata.go
Comment (tigrannajaryan) [inline receiver/k8sclusterreceiver/collection/metadata.go:131]: Is it also needed to check for removed keys, i.e. keys that exist in `old` but do not exist in `new` and generate `KubernetesMetadataUpdate` for such keys?

File: receiver/k8sclusterreceiver/collection/metadata.go
Comment (tigrannajaryan) [inline receiver/k8sclusterreceiver/collection/metadata.go:73]: Can we define a type alias for the string in `map[string]*KubernetesMetadata` to give it a descriptive name and make it clear that it is the resource id?

File: receiver/k8sclusterreceiver/collection/metadata.go
Comment (tigrannajaryan) [inline receiver/k8sclusterreceiver/collection/metadata.go:108]: What is the "key" in these maps? Is it possible to define type alias to make it clear?

File: receiver/k8sclusterreceiver/collection/metadata.go
Comment (asuresh4) [reply receiver/k8sclusterreceiver/collection/metadata.go]: Wrapped these maps around `MetadataDelta`.

File: receiver/k8sclusterreceiver/collection/metadata.go
Comment (asuresh4) [reply receiver/k8sclusterreceiver/collection/metadata.go:122]: In that case we wouldn't be interested in a `KubernetesMetadataUpdate` because the existence of a resource in a previous revision but not in the current one would most likely mean that the resource is no longer active. In such cases, we can skip the resource and still have the last metadata for the resource.

File: receiver/k8sclusterreceiver/collection/metadata.go
Comment (asuresh4) [reply receiver/k8sclusterreceiver/collection/metadata.go]: Added `collection.ResourceID`.

File: receiver/k8sclusterreceiver/collection/metadata.go
Comment (asuresh4) [reply receiver/k8sclusterreceiver/collection/metadata.go]: Added an example in the comments. Let me know if that makes it clear what the keys in these maps are. If not, we could maybe add a type called `metadataKey`?

File: receiver/k8sclusterreceiver/collection/metadata.go
Comment (tigrannajaryan) [inline receiver/k8sclusterreceiver/collection/metadata.go:145]: Interesting use of typo :-)

File: receiver/k8sclusterreceiver/collection/metadata.go
Comment (tigrannajaryan) [inline receiver/k8sclusterreceiver/collection/metadata.go:168]: Would be nice to tell here that nil is returned when old and new are equal and delta is empty.

File: receiver/k8sclusterreceiver/collection/metadata.go
Comment (tigrannajaryan) [reply receiver/k8sclusterreceiver/collection/metadata.go]: Nice examples.

File: receiver/k8sclusterreceiver/README.md
Comment (tigrannajaryan) [inline receiver/k8sclusterreceiver/README.md:61]:
```suggestion
  ResourceID            ResourceID
```

File: receiver/k8sclusterreceiver/README.md
Comment (tigrannajaryan) [inline receiver/k8sclusterreceiver/README.md:64]:
```suggestion
  MetadataDelta
```

**Comment by tigrannajaryan** [comment]
Overall LGTM, only a couple minor comments left.

File: receiver/k8sclusterreceiver/collection/metadata.go
Comment (asuresh4) [reply receiver/k8sclusterreceiver/collection/metadata.go]: Done.

File: receiver/k8sclusterreceiver/README.md
Comment (asuresh4) [reply receiver/k8sclusterreceiver/README.md]: Updated.

File: receiver/k8sclusterreceiver/README.md
Comment (asuresh4) [reply receiver/k8sclusterreceiver/README.md]: Updated.

**Review by tigrannajaryan — APPROVED** [review]

**Comment by tigrannajaryan** [comment]
Keeping open for other approvers to have a look. Will merge tomorrow if there are no other comments.

**Pull request merged**
